File: src/web-app/src/lib/archived-projects.ts
import { normalizeWorkdirKey, workdirDisplay } from "./workdir-display"

/** Minimal shape needed to derive a project — a subset of the store's ArchivedSession. */
export interface ArchivedLike {
  workdir: string
  repo_root?: string
  killed_at?: string
}

export interface ArchivedProject {
  /** normalizeWorkdirKey result — identity for dedupe + filtering. */
  key: string
  /** Display label: shortened path with parent folder. */
  label: string
  /** Number of archived sessions in this project. */
  count: number
}

/** A session's project path: its repo (for worktrees) else its workdir. */
function projectPath(s: ArchivedLike): string {
  return s.repo_root ?? s.workdir
}

/**
 * Shortened path with parent folder — delegates to the shared workdir label so
 * archived projects, session groups, and chat headers all format identically.
 */
export function projectLabel(workdir: string, homeDir?: string | null): string {
  return workdirDisplay(workdir, homeDir).label
}

/** Distinct projects across archived sessions, most-recently-archived first. */
export function archivedProjects(sessions: ArchivedLike[], homeDir?: string | null): ArchivedProject[] {
  const byKey = new Map<string, { key: string; label: string; count: number; latest: string }>()
  for (const s of sessions) {
    const path = projectPath(s)
    const key = normalizeWorkdirKey(path, homeDir)
    const killed = s.killed_at ?? ""
    const existing = byKey.get(key)
    if (existing) {
      existing.count += 1
      if (killed > existing.latest) existing.latest = killed
    } else {
      byKey.set(key, { key, label: projectLabel(path, homeDir), count: 1, latest: killed })
    }
  }
  return [...byKey.values()]
    .sort((a, b) => (a.latest === b.latest ? a.label.localeCompare(b.label) : b.latest.localeCompare(a.latest)))
    .map(({ key, label, count }) => ({ key, label, count }))
}

/** Sessions in the given project (by key). A null/empty key returns all sessions. */
export function filterByProject<T extends ArchivedLike>(
  sessions: T[],
  key: string | null,
  homeDir?: string | null,
): T[] {
  if (!key) return sessions
  return sessions.filter((s) => normalizeWorkdirKey(projectPath(s), homeDir) === key)
}

File: apps/shared/src/commonMain/kotlin/dev/supermux/session/ArchivedProjects.kt
package dev.supermux.session

import dev.supermux.net.ArchivedDto

/** A distinct project across archived sessions. [key] is repo_root ?: workdir. */
data class ArchivedProject(val key: String, val label: String, val count: Int)

/** A session's project key: its repo (for worktrees) else its workdir — matches groupSessions. */
private fun projectKey(s: ArchivedDto): String = s.repo_root ?: s.workdir

/**
 * Distinct projects across archived sessions, most-recently-archived first.
 * Label uses [formatWorkdir]; ties broken alphabetically by label.
 */
fun archivedProjects(sessions: List<ArchivedDto>, home: String?): List<ArchivedProject> {
    data class Acc(val key: String, val label: String, var count: Int, var latest: String)
    val byKey = LinkedHashMap<String, Acc>()
    for (s in sessions) {
        val key = projectKey(s)
        val killed = s.killed_at ?: ""
        val acc = byKey[key]
        if (acc != null) {
            acc.count += 1
            if (killed > acc.latest) acc.latest = killed
        } else {
            byKey[key] = Acc(key, formatWorkdir(key, home), 1, killed)
        }
    }
    return byKey.values
        .sortedWith(compareByDescending<Acc> { it.latest }.thenBy { it.label })
        .map { ArchivedProject(it.key, it.label, it.count) }
}

/** Sessions in the given project (by key). A null key returns all sessions. */
fun filterArchivedByProject(sessions: List<ArchivedDto>, key: String?): List<ArchivedDto> =
    if (key == null) sessions else sessions.filter { projectKey(it) == key }
